CMakeLists.txt

```diff
@@ -76,46 +80,72 @@ if (NOT CMAKE_BUILD_TYPE)
+#    (+ VERBOSE_LLVM_DOWNLOAD: Shows download progress (Default=OFF))
+#  * VERONA_USE_EXTERNAL_LLVM_DIR: The path of a pre-built install dir of an LLVM toolchain.
+#  * VERONA_BUILD_LLVM_SUBMODULE: Builds from the submodule.
+#    (+ LLVM_EXTRA_CMAKE_ARGS: Extra CMake arguments to build LLVM)
+#
+# Note: On both download and build options, the LLVM checkout are guaranteed to
+# be the same as the submodule, and have been validated by CI. For the external
+# build, users need to make sure they either get the same commit or they know what
+# they're doing.
+if (VERONA_USE_EXTERNAL_LLVM_DIR OR (VERONA_BUILD_LLVM_SUBMODULE OR LLVM_EXTRA_CMAKE_ARGS))
+  if (VERONA_USE_EXTERNAL_LLVM_DIR AND (VERONA_BUILD_LLVM_SUBMODULE OR LLVM_EXTRA_CMAKE_ARGS))
+    message (SEND_ERROR "Incompatible options VERONA_BUILD_LLVM_SUBMODULE/LLVM_EXTRA_CMAKE_ARGS and VERONA_USE_EXTERNAL_LLVM_DIR.")
+  elseif (LLVM_EXTRA_CMAKE_ARGS AND NOT VERONA_BUILD_LLVM_SUBMODULE)
+    message (SEND_ERROR "Invalid use of LLVM_EXTRA_CMAKE_ARGS without VERONA_BUILD_LLVM_SUBMODULE.")
+  else()
+    if (VERONA_USE_EXTERNAL_LLVM_DIR)
+      message (STATUS "Import LLVM from: ${VERONA_USE_EXTERNAL_LLVM_DIR}")
+    elseif(VERONA_BUILD_LLVM_SUBMODULE)
+      if (LLVM_EXTRA_CMAKE_ARGS)
+        message (STATUS "Build LLVM with args: ${LLVM_EXTRA_CMAKE_ARGS}")
+      else()
+        message (STATUS "Build LLVM: ${VERONA_BUILD_LLVM_SUBMODULE}")
+      endif()
+    endif()
+  endif ()
+  # Download is default, so if we pick one of these, we want to turn off download
+  set(VERONA_DOWNLOAD_LLVM OFF)
+else()
+  set(VERONA_DOWNLOAD_LLVM ON)
+  message (STATUS "Download LLVM: ${VERONA_DOWNLOAD_LLVM}")
+endif()
 
 ##########################################################
-#  Configure MLIR installation
+#  Configure LLVM installation
 ##########################################################
-if (VERONA_EXTERNAL_LLVM_DIR)
-  # Fake target for MLIR, it is built externally.
+
+if(VERONA_USE_EXTERNAL_LLVM_DIR)
+  # Fake target for LLVM, it is built externally.
   add_custom_target(external)
   # Use existing installation from system
   list (APPEND VERONA_EXTRA_CMAKE_ARGS
-    -DVERONA_LLVM_LOCATION=${VERONA_EXTERNAL_LLVM_DIR}
+    -DVERONA_LLVM_LOCATION=${VERONA_USE_EXTERNAL_LLVM_DIR}
   )
 else()
-  set (MLIR_INSTALL ${CMAKE_BINARY_DIR}/$<CONFIG>/mlir)
-  #  Build or Download MLIR installation
+  set(LLVM_INSTALL ${CMAKE_BINARY_DIR}/$<CONFIG>/mlir)
+  #  Build or Download LLVM installation
   set (EXTERNAL_EXTRA_CMAKE_ARGS)
   list (APPEND EXTERNAL_EXTRA_CMAKE_ARGS 
     -DVERONA_DOWNLOAD_LLVM=${VERONA_DOWNLOAD_LLVM}
-    -DMLIR_INSTALL=${MLIR_INSTALL}
+    -DLLVM_INSTALL=${LLVM_INSTALL}
     -DCMAKE_BUILD_TYPE=$<CONFIG>)
 
-  if (NOT VERONA_DOWNLOAD_LLVM)
+  # Extra options for building or downloading LLVM
+  if(VERONA_BUILD_LLVM_SUBMODULE)
     list (APPEND EXTERNAL_EXTRA_CMAKE_ARGS
       -DLLVM_EXTRA_CMAKE_ARGS=${LLVM_EXTRA_CMAKE_ARGS}
     )
-  else ()
+  else(VERONA_DOWNLOAD_LLVM)
```

Comment:
Should this not be elseif? From the CMake documentation `Per legacy, the else() and endif() commands admit an optional <condition> argument. If used, it must be a verbatim repeat of the argument of the opening if command.`

Reply:
It should. And I thought I had fixed that... :(

Reply:
Fixed it. Thanks for checking!